File: LittleFencer-iOS/LittleFencer/Logic/Detectors/AdvanceDetector.swift
import CoreGraphics
import Foundation

/// Detects the Advance (前进步) footwork
///
/// Biomechanics:
/// - Front foot lifts heel-first, moves forward, lands heel-first
/// - Back foot pushes off, then lands
/// - Fencer returns to En Garde stance
/// - Movement accelerates (slow start, quick finish)
class AdvanceDetector: ActionDetector {
    
    let targetAction = SaberAction.advance
    
    private var isAdvancing = false
    private var advanceStartTime: Double = 0
    private var startPosition: CGFloat = 0
    
    // MARK: - Thresholds
    
    private let MIN_ADVANCE_DISTANCE: CGFloat = 0.05
    private let MAX_ADVANCE_DURATION: Double = 800
    private let MOVEMENT_VELOCITY_THRESHOLD: CGFloat = 0.02
    private let KNEE_MIN: CGFloat = 80
    private let KNEE_MAX: CGFloat = 140
    
    func detect(current: PoseFrame, history: [PoseFrame]) -> ActionResult {
        guard history.count >= 5, current.landmarks.count >= 33 else {
            return .none
        }
        
        let landmarks = current.landmarks
        
        // Extract key points
        let leftAnkle = landmarks[PoseLandmarks.LEFT_ANKLE]
        let rightAnkle = landmarks[PoseLandmarks.RIGHT_ANKLE]
        let leftHip = landmarks[PoseLandmarks.LEFT_HIP]
        let rightHip = landmarks[PoseLandmarks.RIGHT_HIP]
        let leftKnee = landmarks[PoseLandmarks.LEFT_KNEE]
        let rightKnee = landmarks[PoseLandmarks.RIGHT_KNEE]
        
        // Determine front/back based on ankle position
        let isLeftFront = leftAnkle.x > rightAnkle.x
        let frontAnkle = isLeftFront ? leftAnkle : rightAnkle
        
        // Calculate hip midpoint (center of mass proxy)
        let hipMidpoint = GeometryUtils.midpoint(leftHip, rightHip)
        
        // Get previous frame for velocity
        guard let prevFrame = history.last,
              prevFrame.landmarks.count >= 33 else {
            return .none
        }
        
        let prevLeftHip = prevFrame.landmarks[PoseLandmarks.LEFT_HIP]
        let prevRightHip = prevFrame.landmarks[PoseLandmarks.RIGHT_HIP]
        let prevHipMidpoint = GeometryUtils.midpoint(prevLeftHip, prevRightHip)
        
        // Calculate forward velocity (positive X = forward in mirror view)
        let deltaTime = current.timeDeltaMs(prevFrame)
        let forwardVelocity: CGFloat = deltaTime > 0 
            ? (hipMidpoint.x - prevHipMidpoint.x) / CGFloat(deltaTime) * 1000 
            : 0
        
        // Calculate knee angle
        let frontKneeAngle: CGFloat
        if isLeftFront {
            frontKneeAngle = GeometryUtils.angleBetweenPoints(leftHip, leftKnee, leftAnkle)
        } else {
            frontKneeAngle = GeometryUtils.angleBetweenPoints(rightHip, rightKnee, rightAnkle)
        }
        
        // State machine
        if !isAdvancing {
            // Check for advance initiation
            if forwardVelocity > MOVEMENT_VELOCITY_THRESHOLD &&
               frontKneeAngle >= KNEE_MIN && frontKneeAngle <= KNEE_MAX {
                isAdvancing = true
                advanceStartTime = current.timestampMs
                startPosition = frontAnkle.x
                return .inProgress(action: .advance, confidence: 0.5)
            }
            return .none
        } else {
            // Advance in progress
            let elapsed = current.timestampMs - advanceStartTime
            let distanceTraveled = frontAnkle.x - startPosition
            
            // Check timeout
            if elapsed > MAX_ADVANCE_DURATION {
                reset()
                return .none
            }
            
            // Check completion
            if abs(forwardVelocity) < MOVEMENT_VELOCITY_THRESHOLD * 0.5 &&
               distanceTraveled > MIN_ADVANCE_DISTANCE {
                
                let quality = evaluateQuality(kneeAngle: frontKneeAngle, 
                                             distance: distanceTraveled, 
                                             durationMs: elapsed)
                let feedback = generateFeedback(kneeAngle: frontKneeAngle)
                
                reset()
                return .completed(action: .advance, quality: quality, 
                                 feedback: feedback, durationMs: elapsed)
            } else if distanceTraveled < -0.02 {
                // Moving backward - abort
                reset()
                return .none
            }
            
            let confidence = min(0.9, max(0.5, distanceTraveled / MIN_ADVANCE_DISTANCE))
            return .inProgress(action: .advance, confidence: confidence)
        }
    }
    
    private func evaluateQuality(kneeAngle: CGFloat, distance: CGFloat, durationMs: Double) -> ActionQuality {
        let kneeGood = kneeAngle >= 90 && kneeAngle <= 120
        let distanceGood = distance > MIN_ADVANCE_DISTANCE * 1.5
        let speedGood = durationMs < 500
        
        if kneeGood && distanceGood && speedGood {
            return .perfect
        } else if kneeGood && (distanceGood || speedGood) {
            return .good
        } else if kneeGood || distanceGood {
            return .acceptable
        }
        return .poor
    }
    
    private func generateFeedback(kneeAngle: CGFloat) -> String? {
        if kneeAngle < 90 { return "Too low!" }
        if kneeAngle > 130 { return "Bend more!" }
        return nil
    }
    
    func reset() {
        isAdvancing = false
        advanceStartTime = 0
        startPosition = 0
    }
}

File: LittleFencer-Android/app/src/main/java/com/littlefencer/app/logic/detectors/AdvanceDetector.kt
package com.littlefencer.app.logic

import android.graphics.PointF
import com.littlefencer.app.utils.GeometryUtils
import com.littlefencer.app.utils.MediaPipeLandmarks

/**
 * AdvanceDetector - Detects the Advance (前进步) footwork.
 * 
 * Biomechanics:
 * - Front foot lifts heel-first, moves forward, lands heel-first
 * - Back foot pushes off, then lands
 * - Fencer returns to En Garde stance
 * - Movement accelerates (slow start, quick finish)
 * 
 * Detection Strategy:
 * 1. Track ankle X positions over time
 * 2. Detect forward translation (both ankles move +X)
 * 3. Verify En Garde stance maintained during movement
 * 4. Confirm return to stable En Garde at end
 */
class AdvanceDetector : ActionDetector {
    
    override val targetAction = SaberAction.ADVANCE
    
    private var isAdvancing = false
    private var advanceStartTime = 0L
    private var startPosition = 0f  // Initial front ankle X
    
    companion object {
        // Minimum distance traveled (normalized)
        private const val MIN_ADVANCE_DISTANCE = 0.05f
        
        // Maximum time for an advance (ms)
        private const val MAX_ADVANCE_DURATION = 800L
        
        // Velocity threshold to detect movement start
        private const val MOVEMENT_VELOCITY_THRESHOLD = 0.02f
        
        // Knee angle range during advance (should maintain En Garde)
        private const val KNEE_MIN = 80f
        private const val KNEE_MAX = 140f
    }
    
    override fun detect(current: PoseFrame, history: List<PoseFrame>): ActionResult {
        if (history.size < 5) return ActionResult.None
        
        val currentLandmarks = current.landmarks
        if (currentLandmarks.size < 33) return ActionResult.None
        
        // Extract key points
        val leftAnkle = currentLandmarks[MediaPipeLandmarks.LEFT_ANKLE]
        val rightAnkle = currentLandmarks[MediaPipeLandmarks.RIGHT_ANKLE]
        val leftHip = currentLandmarks[MediaPipeLandmarks.LEFT_HIP]
        val rightHip = currentLandmarks[MediaPipeLandmarks.RIGHT_HIP]
        val leftKnee = currentLandmarks[MediaPipeLandmarks.LEFT_KNEE]
        val rightKnee = currentLandmarks[MediaPipeLandmarks.RIGHT_KNEE]
        
        // Determine front/back based on ankle position
        val isLeftFront = leftAnkle.x > rightAnkle.x
        val frontAnkle = if (isLeftFront) leftAnkle else rightAnkle
        val backAnkle = if (isLeftFront) rightAnkle else leftAnkle
        
        // Calculate hip midpoint (center of mass proxy)
        val hipMidpoint = GeometryUtils.midpoint(leftHip, rightHip)
        
        // Get previous hip midpoint for velocity
        val prevFrame = history.lastOrNull() ?: return ActionResult.None
        val prevLandmarks = prevFrame.landmarks
        if (prevLandmarks.size < 33) return ActionResult.None
        
        val prevLeftHip = prevLandmarks[MediaPipeLandmarks.LEFT_HIP]
        val prevRightHip = prevLandmarks[MediaPipeLandmarks.RIGHT_HIP]
        val prevHipMidpoint = GeometryUtils.midpoint(prevLeftHip, prevRightHip)
        
        // Calculate forward velocity (positive X = forward in mirror view)
        val deltaTime = current.timeDeltaMs(prevFrame)
        val forwardVelocity = if (deltaTime > 0) {
            (hipMidpoint.x - prevHipMidpoint.x) / deltaTime * 1000f
        } else 0f
        
        // Calculate knee angles
        val leftKneeAngle = GeometryUtils.angleBetweenPoints(
            leftHip, leftKnee, currentLandmarks[MediaPipeLandmarks.LEFT_ANKLE]
        )
        val rightKneeAngle = GeometryUtils.angleBetweenPoints(
            rightHip, rightKnee, currentLandmarks[MediaPipeLandmarks.RIGHT_ANKLE]
        )
        val frontKneeAngle = if (isLeftFront) leftKneeAngle else rightKneeAngle
        
        // State machine
        return if (!isAdvancing) {
            // Check for advance initiation
            if (forwardVelocity > MOVEMENT_VELOCITY_THRESHOLD && 
                frontKneeAngle in KNEE_MIN..KNEE_MAX) {
                // Start tracking advance
                isAdvancing = true
                advanceStartTime = current.timestampMs
                startPosition = frontAnkle.x
                
                ActionResult.InProgress(
                    action = SaberAction.ADVANCE,
                    confidence = 0.5f
                )
            } else {
                ActionResult.None
            }
        } else {
            // Advance in progress - check completion or failure
            val elapsed = current.timestampMs - advanceStartTime
            val distanceTraveled = frontAnkle.x - startPosition
            
            // Check timeout
            if (elapsed > MAX_ADVANCE_DURATION) {
                reset()
                return ActionResult.None
            }
            
            // Check if movement has stopped (velocity near zero) and distance met
            if (kotlin.math.abs(forwardVelocity) < MOVEMENT_VELOCITY_THRESHOLD * 0.5f &&
                distanceTraveled > MIN_ADVANCE_DISTANCE) {
                
                // Evaluate quality
                val quality = evaluateQuality(frontKneeAngle, distanceTraveled, elapsed)
                val feedback = generateFeedback(frontKneeAngle)
                
                reset()
                
                ActionResult.Completed(
                    action = SaberAction.ADVANCE,
                    quality = quality,
                    feedback = feedback,
                    durationMs = elapsed
                )
            } else if (distanceTraveled < -0.02f) {
                // Moving backward - abort
                reset()
                ActionResult.None
            } else {
                ActionResult.InProgress(
                    action = SaberAction.ADVANCE,
                    confidence = (distanceTraveled / MIN_ADVANCE_DISTANCE).coerceIn(0.5f, 0.9f)
                )
            }
        }
    }
    
    private fun evaluateQuality(kneeAngle: Float, distance: Float, durationMs: Long): ActionQuality {
        val kneeGood = kneeAngle in 90f..120f
        val distanceGood = distance > MIN_ADVANCE_DISTANCE * 1.5f
        val speedGood = durationMs < 500L
        
        return when {
            kneeGood && distanceGood && speedGood -> ActionQuality.PERFECT
            kneeGood && (distanceGood || speedGood) -> ActionQuality.GOOD
            kneeGood || distanceGood -> ActionQuality.ACCEPTABLE
            else -> ActionQuality.POOR
        }
    }
    
    private fun generateFeedback(kneeAngle: Float): String? {
        return when {
            kneeAngle < 90f -> "Too low!"
            kneeAngle > 130f -> "Bend more!"
            else -> null
        }
    }
    
    override fun reset() {
        isAdvancing = false
        advanceStartTime = 0L
        startPosition = 0f
    }
}
